server/channels/app/post_persistent_notification.go

```diff
@@ -0,0 +1,367 @@
+	channelGroupMap := make(map[string]map[string]*model.Group, len(channelsMap))
+	channelProfileMap := make(map[string]model.UserMap, len(channelsMap))
+	channelKeywords := make(map[string]map[string][]string, len(channelsMap))
+	channelNotifyProps := make(map[string]map[string]model.StringMap, len(channelsMap))
+	for _, c := range channelsMap {
+		// In DM, notifications can't be send to any 3rd person.
+		if c.Type != model.ChannelTypeDirect {
+			groups, err := a.getGroupsAllowedForReferenceInChannel(c, teamsMap[c.TeamId])
+			if err != nil {
+				return nil, nil, nil, nil, errors.Wrapf(err, "failed to get profiles for channel %s", c.Id)
+			}
+			channelGroupMap[c.Id] = make(map[string]*model.Group, len(groups))
+			for k, v := range groups {
+				channelGroupMap[c.Id][k] = v
+			}
+			props, err := a.Srv().Store().Channel().GetAllChannelMembersNotifyPropsForChannel(c.Id, true)
+			if err != nil {
+				return nil, nil, nil, nil, errors.Wrapf(err, "failed to get profiles for channel %s", c.Id)
+			}
+			channelNotifyProps[c.Id] = props
+		}
+
+		profileMap, err := a.Srv().Store().User().GetAllProfilesInChannel(context.Background(), c.Id, true)
+		if err != nil {
+			return nil, nil, nil, nil, errors.Wrapf(err, "failed to get profiles for channel %s", c.Id)
+		}
+
+		channelKeywords[c.Id] = make(map[string][]string, len(profileMap))
+		validProfileMap := make(map[string]*model.User, len(profileMap))
+		for k, v := range profileMap {
+			if v.IsBot {
+				continue
+			}
+			validProfileMap[k] = v
+			channelKeywords[c.Id]["@"+v.Username] = []string{k}
+		}
+		channelProfileMap[c.Id] = validProfileMap
+	}
+	return channelGroupMap, channelProfileMap, channelKeywords, channelNotifyProps, nil
+}
+
+func (a *App) channelTeamMapsForPosts(posts []*model.Post) (map[string]*model.Channel, map[string]*model.Team, error) {
+	channelIds := make(model.StringSet)
+	for _, p := range posts {
+		channelIds.Add(p.ChannelId)
+	}
+	channels, err := a.Srv().Store().Channel().GetChannelsByIds(channelIds.Val(), false)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to get teams by IDs")
+	}
+	channelsMap := make(map[string]*model.Channel, len(channels))
+	for _, c := range channels {
+		channelsMap[c.Id] = c
+	}
+
+	teamIds := make(model.StringSet)
+	for _, c := range channels {
+		if c.TeamId != "" {
+			teamIds.Add(c.TeamId)
+		}
+	}
+	teams := make([]*model.Team, 0, len(teamIds))
+	if len(teamIds) > 0 {
+		teams, err = a.Srv().Store().Team().GetMany(teamIds.Val())
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "failed to get teams by IDs")
+		}
+	}
+	teamsMap := make(map[string]*model.Team, len(teams))
+	for _, t := range teams {
+		teamsMap[t.Id] = t
+	}
+	return channelsMap, teamsMap, nil
+}
+
+func (a *App) sendPersistentNotifications(post *model.Post, channel *model.Channel, team *model.Team, mentions *ExplicitMentions, profileMap model.UserMap, channelNotifyProps map[string]map[string]model.StringMap) error {
```

Comment:
The idea was to keep them separate as requirements for sending PNs may differ from a regular notification. ATM, there are slight changes but it could diverge further in future.

Reply:
Requirements or methodology? I can see how the conditions under which the notifications are sent are going to be different, but it's not clear why we need to duplicate the method of actually sending those. At the very least, can we factor out the differences?

Reply:
The actual sending of the notifications is still [common](https://github.com/mattermost/mattermost-server/blob/persist-notifications/server/channels/app/post_persistent_notification.go#L281). It's the people who need to be notified which is different or could be different in future.